### PR TITLE
eip-7251: process_registry_updates

### PR DESCRIPTION
**Pull request by prestonvanloon**
**What type of PR is this?**

Feature

**What does this PR do? Why is it needed?**

Required for Electra.

Extracted from #13903.

**Which issues(s) does this PR fix?**

Tracking @ #13849

**Other notes for review**

### CODE REVIEW

File: beacon-chain/core/electra/registry_updates.go [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+//		            and validator.effective_balance <= EJECTION_BALANCE
+//		        ):
+//		            initiate_validator_exit(state, ValidatorIndex(index))
+//
+//	         # Activate all eligible validators
+//	         activation_epoch = compute_activation_exit_epoch(get_current_epoch(state))
+//	         for validator in state.validators:
+//	             if is_eligible_for_activation(state, validator):
+//	                 validator.activation_epoch = activation_epoch
+func ProcessRegistryUpdates(ctx context.Context, state state.BeaconState) (state.BeaconState, error) {
```
Comment (james-prysm) [inline beacon-chain/core/electra/registry_updates.go:36]: you can add a span for this like your other PRs

File: beacon-chain/core/electra/registry_updates.go [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+	vals := state.Validators()
+	for idx, val := range vals {
+		if helpers.IsEligibleForActivationQueue(val, currentEpoch) {
+			val.ActivationEligibilityEpoch = currentEpoch + 1
+			if err := state.UpdateValidatorAtIndex(primitives.ValidatorIndex(idx), val); err != nil {
+				return nil, err
+			}
+		}
+		if helpers.IsActiveValidator(val, currentEpoch) && val.EffectiveBalance <= ejectionBal {
+			var err error
```
Comment (james-prysm) [inline beacon-chain/core/electra/registry_updates.go:49]: having this var here feels weird , maybe it sjust so you can set the state  on the existing 
on line 51

Comment (prestonvanloon) [reply beacon-chain/core/electra/registry_updates.go:49]: yeah exactly, err is not defined in this scope and i need it for line 51.

I could switch L49 and L50 to move it closer to its use

File: beacon-chain/core/electra/registry_updates.go (hunk 1 above)
Comment (prestonvanloon) [reply beacon-chain/core/electra/registry_updates.go:36]: I dont know that it is necessary yet. Perhaps other spans should be removed unless it is considered a hotspot in processing

**Comment by prestonvanloon** [comment]
Working on fixing this up. Just needs more unit tests (spectests are passing)

File: beacon-chain/core/electra/registry_updates.go [hunk 3]
```diff
@@ -32,34 +33,37 @@ import (
...
+		// Handle validator ejections.
+		if val.EffectiveBalance <= ejectionBal && helpers.IsActiveValidator(val, currentEpoch) {
 			var err error
-			maxExitEpoch, churn := validators.MaxExitEpochAndChurn(state) 
-			state, _, err = validators.InitiateValidatorExit(ctx, state, primitives.ValidatorIndex(idx), maxExitEpoch, churn)
+			// exitQueueEpoch and churn arguments are not used in electra.
+			state, _, err = validators.InitiateValidatorExit(ctx, state, primitives.ValidatorIndex(idx), 0 /*exitQueueEpoch*/, 0 /*churn*/)
 			if err != nil {
-				return nil, err
+				return err
```
Comment (james-prysm) [inline beacon-chain/core/electra/registry_updates.go:55]: maybe worth adding an errors.Wrap here to know it's bubbling up for initiate validator exit

File: beacon-chain/core/electra/registry_updates.go [hunk 4]
```diff
@@ -28,7 +33,37 @@ import (
...
+			val.ActivationEligibilityEpoch = currentEpoch + 1
+			if err := state.UpdateValidatorAtIndex(primitives.ValidatorIndex(idx), val); err != nil {
+				return err
+			}
+		}
+		// Handle validator ejections.
+		if val.EffectiveBalance <= ejectionBal && helpers.IsActiveValidator(val, currentEpoch) {
+			var err error
+			// exitQueueEpoch and churn arguments are not used in electra.
+			state, _, err = validators.InitiateValidatorExit(ctx, state, primitives.ValidatorIndex(idx), 0 /*exitQueueEpoch*/, 0 /*churn*/)
```
Comment (saolyn) [inline beacon-chain/core/electra/registry_updates.go:53]: Are these values supposed to be 0?

Comment (james-prysm) [reply beacon-chain/core/electra/registry_updates.go:53]: in the function it says it's ignored for electra so I guess it can be anything but 0 seems fine

**Review by james-prysm — APPROVED** [review]
lgtm

**Review by saolyn — DISMISSED** [review]

File: beacon-chain/core/electra/registry_updates.go [hunk 5]
```diff
@@ -28,7 +34,37 @@ import (
...
 //	             if is_eligible_for_activation(state, validator):
 //	                 validator.activation_epoch = activation_epoch
-func ProcessRegistryUpdates(ctx context.Context, state state.BeaconState) (state.BeaconState, error) {
-	// TODO: replace with real implementation
-	return state, nil
+func ProcessRegistryUpdates(ctx context.Context, state state.BeaconState) error {
+	currentEpoch := time.CurrentEpoch(state)
+	ejectionBal := params.BeaconConfig().EjectionBalance
+	activationEpoch := helpers.ActivationExitEpoch(currentEpoch)
+	vals := state.Validators()
```
Comment (terencechain) [inline beacon-chain/core/electra/registry_updates.go:41]: `state.Validators()` copies the entire validator set and this is very expensive given the current validator set size. 
Have you thought about using `ValidatorsReadOnly` or `ReadFromEveryValidator`

**Review by terencechain — CHANGES_REQUESTED** [review]

File: beacon-chain/core/electra/registry_updates.go (hunk 5 above)
Comment (prestonvanloon) [reply beacon-chain/core/electra/registry_updates.go:41]: Good point. I'll do something about this

**Comment by prestonvanloon** [comment]
Design in #14130 was approved. I'll update this PR today to adopt that design.

File: beacon-chain/core/electra/registry_updates.go (hunk 5 above)
Comment (prestonvanloon) [reply beacon-chain/core/electra/registry_updates.go]: This is done now. Using the same design as #14130

**Review by terencechain — APPROVED** [review]

**Review by terencechain — APPROVED** [review]

**Review by potuz — APPROVED** [review]

**Pull request merged**